File: dotnet/core/workspace/csharp/adapters/allors.workspace.adapters.remote/session/Session.cs
// <copyright file="RemoteSession.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Remote
{
    using System;
    using System.Linq;
    using System.Threading.Tasks;
    using Allors.Protocol.Json.Api.Invoke;
    using Allors.Protocol.Json.Api.Pull;
    using Allors.Protocol.Json.Api.Push;
    using Allors.Protocol.Json.Api.Sync;
    using Data;
    using Meta;
    using Protocol.Json;
    using InvokeOptions = InvokeOptions;

    public class Session : Adapters.Session
    {
        internal Session(Workspace workspace, ISessionServices sessionServices) : base(workspace, sessionServices)
        {
            this.Services.OnInit(this);
            this.DatabaseConnection = workspace.DatabaseConnection;
        }

        public DatabaseConnection DatabaseConnection { get; set; }

        public new Workspace Workspace => (Workspace)base.Workspace;

        public override async Task<IInvokeResult> Invoke(Method method, InvokeOptions options = null) => await this.Invoke(new[] { method }, options);

        public override async Task<IInvokeResult> Invoke(Method[] methods, InvokeOptions options = null)
        {
            var invokeRequest = new InvokeRequest
            {
                l = methods.Select(v => new Invocation
                {
                    i = v.Object.Id,
                    v = ((Strategy)v.Object.Strategy).DatabaseOriginState.Version,
                    m = v.MethodType.Tag
                }).ToArray(),
                o = options != null
                    ? new Allors.Protocol.Json.Api.Invoke.InvokeOptions
                    {
                        c = options.ContinueOnError,
                        i = options.Isolated
                    }
                    : null
            };

            var invokeResponse = await this.Workspace.DatabaseConnection.Invoke(invokeRequest);
            return new InvokeResult(this, invokeResponse);
        }

        public override async Task<IPullResult> Pull(params Pull[] pulls)
        {
            var pullRequest = new PullRequest { l = pulls.Select(v => v.ToJson(this.DatabaseConnection.UnitConvert)).ToArray() };

            var pullResponse = await this.Workspace.DatabaseConnection.Pull(pullRequest);
            return await this.OnPull(pullResponse);
        }

        public override async Task<IPullResult> Pull(Procedure procedure, params Pull[] pulls)
        {
            var pullRequest = new PullRequest
            {
                p = procedure.ToJson(this.DatabaseConnection.UnitConvert),
                l = pulls.Select(v => v.ToJson(this.DatabaseConnection.UnitConvert)).ToArray()
            };

            var pullResponse = await this.Workspace.DatabaseConnection.Pull(pullRequest);
            return await this.OnPull(pullResponse);
        }

        public override async Task<IPushResult> Push()
        {
            var result = await this.PushToDatabase();

            if (!result.HasErrors)
            {
                this.PushToWorkspace(result);
            }

            return result;
        }

        private async Task<IPullResult> OnPull(PullResponse pullResponse)
        {
            var syncRequest = this.Workspace.DatabaseConnection.OnPullResponse(pullResponse);
            if (syncRequest.o.Length > 0)
            {
                await this.Sync(syncRequest);
            }

            foreach (var v in pullResponse.p)
            {
                if (this.StrategyByWorkspaceId.TryGetValue(v.i, out var strategy))
                {
                    ((DatabaseOriginState)strategy.DatabaseOriginState).OnPulled();
                }
                else
                {
                    this.InstantiateDatabaseStrategy(v.i);
                }
            }

            return new PullResult(this, pullResponse);
        }

        private async Task Sync(SyncRequest syncRequest)
        {
            var database = (DatabaseConnection)base.Workspace.DatabaseConnection;
            var syncResponse = await database.Sync(syncRequest);
            var securityRequest = database.OnSyncResponse(syncResponse);

            foreach (var databaseObject in syncResponse.o)
            {
                if (!this.StrategyByWorkspaceId.TryGetValue(databaseObject.i, out var strategy))
                {
                    this.InstantiateDatabaseStrategy(databaseObject.i);
                }
                else
                {
                    ((DatabaseOriginState)strategy.DatabaseOriginState).OnPulled();
                }
            }

            if (securityRequest != null)
            {
                var securityResponse = await database.Security(securityRequest);
                securityRequest = database.SecurityResponse(securityResponse);

                if (securityRequest != null)
                {
                    securityResponse = await database.Security(securityRequest);
                    database.SecurityResponse(securityResponse);
                }
            }
        }

        public override T Create<T>(IClass @class)
        {
            var workspaceId = base.Workspace.WorkspaceIdGenerator.Next();
            var strategy = new Strategy(this, @class, workspaceId);
            this.AddStrategy(strategy);

            if (@class.Origin != Origin.Session)
            {
                this.PushToWorkspaceTracker.OnCreated(strategy);
                if (@class.Origin == Origin.Database)
                {
                    this.PushToDatabaseTracker.OnCreated(strategy);
                }
            }

            this.ChangeSetTracker.OnCreated(strategy);

            return (T)strategy.Object;
        }

        public override Adapters.Strategy InstantiateDatabaseStrategy(long id)
        {
            var databaseRecord = (DatabaseRecord)base.Workspace.DatabaseConnection.GetRecord(id);
            var strategy = new Strategy(this, databaseRecord);
            this.AddStrategy(strategy);

            this.ChangeSetTracker.OnInstantiated(strategy);

            return strategy;
        }

        protected override Adapters.Strategy InstantiateWorkspaceStrategy(long id)
        {
            if (!base.Workspace.WorkspaceClassByWorkspaceId.TryGetValue(id, out var @class))
            {
                return null;
            }

            var strategy = new Strategy(this, @class, id);
            this.AddStrategy(strategy);

            this.ChangeSetTracker.OnInstantiated(strategy);

            return strategy;
        }

        private PushRequest PushRequest() => new PushRequest
        {
            n = this.PushToDatabaseTracker.Created?.Select(v => ((Strategy)v).DatabasePushNew()).ToArray(),
            o = this.PushToDatabaseTracker.Changed?.Select(v => ((Strategy)v.Strategy).DatabasePushExisting()).ToArray()
        };

        private async Task<PushResult> PushToDatabase()
        {
            var pushRequest = this.PushRequest();
            var pushResponse = await this.Workspace.DatabaseConnection.Push(pushRequest);

            if (pushResponse.HasErrors)
            {
                return new PushResult(this, pushResponse);
            }

            if (pushResponse.n != null)
            {
                foreach (var pushResponseNewObject in pushResponse.n)
                {
                    var workspaceId = pushResponseNewObject.w;
                    var databaseId = pushResponseNewObject.d;

                    this.OnDatabasePushResponseNew(workspaceId, databaseId);
                }
            }

            this.PushToDatabaseTracker.Created = null;

            if (pushRequest.o != null)
            {
                foreach (var id in pushRequest.o.Select(v => v.d))
                {
                    var strategy = this.GetStrategy(id);
                    this.OnDatabasePushResponse(strategy);
                }
            }

            return new PushResult(this, pushResponse);
        }
    }
}

File: typescript/libs/workspace/adapters/json/system/src/lib/runtime/Session.ts
import { InvokeRequest, PullRequest, PullResponse, PushRequest } from '@allors/protocol/json/system';
import { Session as SystemSession } from '@allors/workspace/adapters/system';
import { IInvokeResult, InvokeOptions, IObject, IPullResult, IPushResult, ISessionServices, Method, Procedure, Pull } from '@allors/workspace/domain/system';
import { Class, Origin } from '@allors/workspace/meta/system';
import { procedureToJson, pullToJson } from '../json/toJson';
import { Database } from './Database';
import { DatabaseOriginState } from './DatabaseOriginState';
import { DatabaseRecord } from './DatabaseRecord';
import { InvokeResult } from './invoke/InvokeResult';
import { PullResult } from './pull/PullResult';
import { PushResult } from './push/PushResult';
import { Strategy } from './Strategy';
import { Workspace } from './Workspace';

export class Session extends SystemSession {
  get database(): Database {
    return this.workspace.database as Database;
  }

  constructor(workspace: Workspace, services: ISessionServices) {
    super(workspace, services);

    this.services.onInit(this);
  }

  async invoke(methods: Method[], options: InvokeOptions): Promise<IInvokeResult> {
    const invokeRequest: InvokeRequest = {
      l: methods.map((v) => {
        return {
          i: v.object.id,
          v: (v.object.strategy as Strategy).DatabaseOriginState.version,
          m: v.methodType.tag,
        };
      }),
      o:
        options != null
          ? {
              c: options.continueOnError,
              i: options.isolated,
            }
          : null,
    };

    const response = await this.database.invoke(invokeRequest);
    return new InvokeResult(this, response);
  }

  async pull(pulls: Pull[]): Promise<IPullResult> {
    const pullRequest: PullRequest = {
      l: pulls.map((v) => pullToJson(v)),
    };

    const result = await this.database.pull(pullRequest);
    return this.OnPull(result);
  }

  async call(procedure: Procedure, ...pulls: Pull[]): Promise<IPullResult> {
    const pullRequest: PullRequest = {
      p: procedureToJson(procedure),
      l: pulls.map((v) => pullToJson(v)),
    };

    const response = await this.database.pull(pullRequest);
    return await this.OnPull(response);
  }

  async push(): Promise<IPushResult> {
    const pushRequest: PushRequest = {
      n: [...this.pushToDatabaseTracker.created].map((v) => (v.DatabaseOriginState as DatabaseOriginState).pushNew()),
      o: [...this.pushToDatabaseTracker.changed].map((v) => (v.strategy.DatabaseOriginState as DatabaseOriginState).pushExisting()),
    };

    const pushResponse = await this.database.push(pushRequest);

    const result = new PushResult(this, pushResponse);

    if(result.hasErrors){
      return result;
    }

    if ((pushResponse.n != null)) {
        for (let pushResponseNewObject of pushResponse.n) {
            let workspaceId = pushResponseNewObject.w;
            let databaseId = pushResponseNewObject.d;
            this.onDatabasePushResponseNew(workspaceId, databaseId);
        }
    }

    this.pushToDatabaseTracker.created = null;

    if ((pushRequest.o != null)) {
        for (let id of pushRequest.o.map(v => v.d)) {
            let strategy = this.getStrategy(id);
            this.onDatabasePushResponse(strategy);
        }

    }

    if (!result.hasErrors) {
        this.PushToWorkspace(result);
    }
  }

  create<T extends IObject>(cls: Class): T {
    const workspaceId = this.workspace.database.nextId();
    const strategy = new Strategy(this, cls, workspaceId);
    this.addStrategy(strategy);
    if (cls.origin != Origin.Session) {
      this.pushToWorkspaceTracker.onCreated(strategy);
      if (cls.origin == Origin.Database) {
        this.pushToDatabaseTracker.onCreated(strategy);
      }
    }

    this.changeSetTracker.onCreated(strategy);
    return strategy.object as T;
  }

  instantiateDatabaseStrategy(id: number): Strategy {
    const databaseRecord = this.workspace.database.getRecord(id) as DatabaseRecord;
    const strategy = Strategy.fromDatabaseRecord(this, databaseRecord);
    this.addStrategy(strategy);
    this.changeSetTracker.onInstantiated(strategy);
    return strategy;
  }

  instantiateWorkspaceStrategy(id: number): Strategy {
    if (!this.workspace.workspaceClassByWorkspaceId.has(id)) {
      return null;
    }

    const cls = this.workspace.workspaceClassByWorkspaceId.get(id);
    const strategy = new Strategy(this, cls, id);
    this.addStrategy(strategy);
    this.changeSetTracker.onInstantiated(strategy);
    return strategy;
  }

  private async OnPull(pullResponse: PullResponse): Promise<IPullResult> {
    const syncRequest = (this.workspace.database as Database).onPullResonse(pullResponse);
    if (syncRequest.o.length > 0) {
      const database = this.workspace.database as Database;
      const syncResponse = await database.sync(syncRequest);
      let securityRequest = database.onSyncResponse(syncResponse);
      for (const v of syncResponse.o) {
        if (!this.strategyByWorkspaceId.has(v.i)) {
          this.instantiateDatabaseStrategy(v.i);
        } else {
          const strategy = this.strategyByWorkspaceId.get(v.i);
          strategy.DatabaseOriginState.onPulled();
        }
      }

      if (securityRequest != null) {
        let securityResponse = await database.security(securityRequest);
        securityRequest = database.securityResponse(securityResponse);
        if (securityRequest != null) {
          securityResponse = await database.security(securityRequest);
          database.securityResponse(securityResponse);
        }
      }
    }

    for (const v of pullResponse.p) {
      if (this.strategyByWorkspaceId.has(v.i)) {
        const strategy = this.strategyByWorkspaceId.get(v.i);
        strategy.DatabaseOriginState.onPulled();
      } else {
        this.instantiateDatabaseStrategy(v.i);
      }
    }

    return new PullResult(this, pullResponse);
  }
}
